MAINT: stats.wilcoxon: small improvements/fixes

chrisb83 opened the pull request:
<!-- 
Thanks for contributing a pull request! Please ensure that
your PR satisfies the checklist before submitting:
https://scipy.github.io/devdocs/dev/contributor/development_workflow.html#checklist-before-submitting-a-pr

Also, please name and describe your PR as you would write a
commit message:
https://scipy.github.io/devdocs/dev/contributor/development_workflow.html#writing-the-commit-message.
However, please only include an issue number in the description, not the title,
and please ensure that any code names containing underscores are enclosed in backticks.

Depending on your changes, you can skip CI operations and save time and energy: 
https://scipy.github.io/devdocs/dev/contributor/continuous_integration.html#skipping

Note that we are a team of volunteers; we appreciate your
patience during the review process.

Again, thanks for contributing!
-->

#### Reference issue
see points 1.-4. in https://github.com/scipy/scipy/pull/19770#issuecomment-1880158299

#### What does this implement/fix?
1. ~use continuity correction by default to improve accuracy if `method == "approx"` (now "asymptotic", see 4.)~
2. respect the method parameter if it is explicitly passed by the user. in particular, if exact calculation is requested, perform it and warn in case of ties/zeros, but do not change to approx
3. improve auto method by performing a permutation test if the sample size is small and there are ties/zeros
4. rename `method='approx'` to `method ='asymptotic'` for consistency with closely related tests ([mannwhitneyu](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.mannwhitneyu.html), [kendalltau](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.kendalltau.html), and [page_trend_test](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.page_trend_test.html)). ~Raise deprecation warning if `approx` is used~

#### Additional information

- The values in the documentation need to be adjusted once there is an agreement about the changes
- The parameters of the PermutationMethod still need to be determined if `method ==  "auto"` (e.g., sample size to achieve sufficient accuracy). One open question is if the results need to be reproducible (currently no seed it set)

chrisb83 committed:
MAINT: use continuity correction in wilcoxon

chrisb83 committed:
MAINT: approx -> asymptotic in wilcoxon

chrisb83 committed:
MAINT: use PermutationTest if ties/zeros and auto method in wilcoxon test

chrisb83 committed:
MAINT: add deprecation warning to wilcoxon test

chrisb83 commented on scipy/stats/_morestats.py:
```diff
@@ -3679,6 +3679,12 @@ def wilcoxon(x, y=None, zero_method="wilcox", correction=False,
     """
+    # replace approx by asymptotic to ensure backwards compatability
+    if method == "approx":
+        depr_msg = ("The method `approx` has been renamed to `aymptotic`. "
```
typo "aymptotic"

chrisb83 commented on scipy/stats/_wilcoxon.py:
```diff
@@ -97,38 +97,45 @@ def _wilcoxon_iv(x, y, zero_method, correction, alternative, method, axis):
         if method not in methods:
             raise ValueError(message)
-    output_z = True if method == 'approx' else False
+    output_z = True if method == 'asymptotic' else False
```
what I noted: if method=auto and we switch to asymptotic later, we do not output z. I am not sure if this was intended

mdhaber replied:
Yes, this is intended. When there are NaNs and `nan_policy='omit'`, the number elements in each slice may vary. Consequently, the exact method may be used for some slices and the asymptotic approximation for others. Returning `z` for some but not others causes problems. (One could argue that `z` _should_ always be included in the result object, but that is a different question.)

mdhaber commented on scipy/stats/_wilcoxon.py:
```diff
@@ -197,18 +206,26 @@ def _wilcoxon_nd(x, y=None, zero_method='wilcox', correction=True,
+    # the decision was to use an exact test, we override this to
+    # a permutation test now (since method=`exact`` is not exact in the
+    # presence of ties)
+    if m_auto and has_ties and method == 'exact':
```
```suggestion
    if (has_zeros or has_ties) and method == 'auto':
```
Why not leave the `method` unresolved until here (instead of changing it and changing it again based on a variable with a name that is less obvious than `has_zeros`)?

chrisb83 replied:
true, that is better. i made a suggestion that resolves to exact or PermutationMethod later.

mdhaber reviewed (commented):
I am going to guess that we won't get much feedback from the mailing list; instead, when people notice that the p-value is non-deterministic, they will complain. 

Perhaps instead we should only let "auto" resolve to a `PermutationMethod` if `n` is small enough to perform an exact permutation test (deterministic). Otherwise, emit a warning suggesting that the user specify a `PermutationTest` explicitly?

mdhaber commented on scipy/stats/tests/test_morestats.py:
```diff
@@ -1658,22 +1674,26 @@ def test_exact_p_1(self, x):
-            w, p = stats.wilcoxon(d)
-        assert_equal(stats.wilcoxon(d, mode="approx"), (w, p))
+        w, p = stats.wilcoxon(d)
+        assert_equal(stats.wilcoxon(d, method=pm), (w, p))
```
These are `np.testing.assert_equal`, so it doesn't matter like it does for `np.testing.assert_allclose` or `xp_assert_equal`, but please order the arguments with the result being tested first and the reference second. ```suggestion
        assert_equal((w, p), stats.wilcoxon(d, method=pm))
```

mdhaber commented on scipy/stats/_morestats.py:
```diff
@@ -3581,7 +3581,7 @@ def wilcoxon(x, y=None, zero_method="wilcox", correction=False,
     The presence of "ties" (i.e. not all elements of ``d`` are unique) or
```
The bit above looks like it needs to be updated since `method='auto'` now selects `PermutationMethod` sometimes.

chrisb83 commented:
**Comparison of `correction=True` vs `correction=False`:**

- simulation of `x` and `y` of random length between 2 and 50
- vary `alternative` (`less` and `two-sided`, `greater` is not needed by symmetry ) and `zero_method` 
- compare p-value of exact and asymptotic method w/ and w/o correction
- overall 20000 test cases

Overall, the p-value are closer to the exact p-value in ~83% of the cases. For larger samples, the improvement tends to be a bit lower:

![image](https://github.com/user-attachments/assets/1fb9fb11-009e-4af8-851c-6e3d5b043885)

The `zero_method` does not have much influence on the results:
![image](https://github.com/user-attachments/assets/6ce30657-4aca-4907-b1fc-fc30714bc561)

If `alternative="two-sided"`, there is an improvement in about 95% of the cases, if `alternative="less"`, it is in about 72% of the cases.

![image](https://github.com/user-attachments/assets/cbd86b7e-1807-45c2-8219-f360ce18c114)

Source code:

```
from scipy import stats
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt


rng = np.random.default_rng(988970986237)

def compare_correction(rng, n_test, alternative, zero_method):
    res = np.zeros((n_test, 4))
    for i in range(n_test):
        N = rng.integers(2, 51)
        x = rng.standard_normal(size=N)
        y = rng.standard_normal(size=N)
        kwargs = {'alternative': alt, 'zero_method': zero_method}
        p_exact = stats.wilcoxon(x, y, method="exact", **kwargs).pvalue
        p_asymp_corr = stats.wilcoxon(x, y, method="approx", correction=True, **kwargs).pvalue
        p_asymp_no_corr = stats.wilcoxon(x, y, method="approx", correction=False, **kwargs).pvalue

        res[i, :] = (N, p_exact, p_asymp_corr, p_asymp_no_corr)

    df = pd.DataFrame(res)
    df.columns = ["N", "p_exact", "p_corr", "p_no_corr"]
    df["delta_corr"] = np.abs(df["p_exact"] - df["p_corr"])
    df["delta_no_corr"] = np.abs(df["p_exact"] - df["p_no_corr"])
    df["corr_better"] = (df["delta_corr"] <= df["delta_no_corr"]).astype(int)
    df["zero_method"] = zero_method
    df["alternative"] = alternative

    return df

df_lst = []
for alt in ["two-sided", "less"]:
    for zm in ["pratt", "wilcox", "zsplit"]:
        df = compare_correction(rng, 2000, alt, zm)
        df_lst.append(df)

df = pd.concat(df_lst)
df.reset_index(inplace=True)

print("Overall stats...")
print(df.describe())
# improvement in 83.5% of all cases with correction=True

print("Stats when correction=True is better...")
print(df.loc[df["corr_better"] == 1, :].describe())
# if correction=True is better, the avg error compared to exact p-value is ~3.5x lower when
# correction=True vs correction=False

print("Stats when correction=True is worse...")
print(df.loc[df["corr_better"] == 0, :].describe())
# if correction=True is worse, the avg error compared to exact p-value is ~2x higher when
# correction=True vs correction=False

grouped = df.groupby(['N', 'corr_better']).size().unstack(fill_value=0)

plt.figure(figsize=(10, 6))
plt.plot(grouped.index, grouped[0], label='no correction better', marker='o')
plt.plot(grouped.index, grouped[1], label='correction better', marker='o')
plt.xlabel('N')
plt.ylabel('Count')
plt.legend(loc="best")
plt.grid(True)
plt.show()


df.groupby(["zero_method", "alternative", "corr_better"])["N"].count()

grouped = df.groupby(['N', 'corr_better', 'zero_method']).size().unstack(['corr_better', 'zero_method'], fill_value=0)

# Plot
plt.figure(figsize=(10, 6))

# Assuming zero_method has unique values you want to compare
for zero_method_value in grouped.columns.levels[1]:
    plt.plot(grouped.index, grouped[(0, zero_method_value)], label=f'no correction better, {zero_method_value}', marker='o')
    plt.plot(grouped.index, grouped[(1, zero_method_value)], label=f'correction better, {zero_method_value}', marker='o')

# Add labels and title
plt.xlabel('N')
plt.ylabel('Count')
plt.title('Count of corr_better = 0 and 1 for each N by zero_method')
plt.legend()
plt.grid(True)
plt.show()

grouped = df.groupby(['N', 'corr_better', 'alternative']).size().unstack(['corr_better', 'alternative'], fill_value=0)

# Plot
plt.figure(figsize=(10, 6))

# Assuming zero_method has unique values you want to compare
for alternative_value in grouped.columns.levels[1]:
    plt.plot(grouped.index, grouped[(0, alternative_value)], label=f'correction better, {alternative_value}', marker='o')
    plt.plot(grouped.index, grouped[(1, alternative_value)], label=f'no correction better, {alternative_value}', marker='o')

# Add labels and title
plt.xlabel('N')
plt.ylabel('Count')
plt.title('Count of corr_better = 0 and 1 for each N by alternative')
plt.legend()
plt.grid(True)
plt.show()
```

chrisb83 commented:
> Perhaps instead we should only let "auto" resolve to a PermutationMethod if n is small enough to perform an exact permutation test (deterministic). Otherwise, emit a warning suggesting that the user specify a PermutationTest explicitly?

I think that makes sense, let me implement it

chrisb83 commented on scipy/stats/_morestats.py:
```diff
@@ -3512,7 +3512,7 @@ def wilcoxon(x, y=None, zero_method="wilcox", correction=False,
         If True, apply continuity correction by adjusting the Wilcoxon rank
         statistic by 0.5 towards the mean value when computing the
-        z-statistic if a normal approximation is used.  Default is False.
+        z-statistic if a normal approximation is used.  Default is True.
```
Done, see below

chrisb83 committed:
MAINT: address review feedback on stats.wilcoxon

[skip circle]

mdhaber commented:
Thanks for the simulations. I also think the plot of the exact and approximate null distributions is enlightening:

<details>

```python3
import numpy as np
import matplotlib.pyplot as plt

from scipy.stats._wilcoxon import WilcoxonDistribution
from scipy import stats

# Compute location and scale of (normal) null distribution
# according to asymptotic approximation
count = 10
mn = count * (count + 1.) * 0.25
se = count * (count + 1.) * (2. * count + 1.)
se = np.sqrt(se / 24)

# Exact null distribution
X = WilcoxonDistribution(count)

# Statistic values near typical significance threshold
k = np.arange(X._dists[count].size)
k = k[X.sf(k) < 0.2]

# Uncorrected/Corrected null distribution
Y = stats.norm(mn, se)
Z = stats.norm(mn + 0.5, se)

# Compute survival function (i.e. p-value with alternative='greater')
p0 = X.sf(k)
py = Y.sf(k)
pz = Z.sf(k)

# Plot results
plt.plot(k, p0, '*', label='exact')
plt.plot(k, py, label='asymptotic, uncorrected')
plt.plot(k, pz, label='asymptotic, corrected')
plt.legend()
plt.xlabel('statistic')
plt.ylabel('survival function')
plt.title('`wilcoxon` null distribution survival function ($n=10$)')
```

</details>
<img width="1181" alt="image" src="https://github.com/user-attachments/assets/6cec125d-db05-4803-81ca-b4d2c5cdb03d">

The uncorrected asymptotic null distribution tends to report lower p-values than the exact null distribution, leading to false discoveries. The correction shifts the null distribution to the right by half an integer to account for the fact that the normal distribution is continuous whereas the exact distribution is discrete, and this substantantially improves the agreement between the asymptotic and exact curves.

We can plot the maximum disagreement between the asymptotic and exact p-values in this domain as a function of the sample size:

<details>

```
err_corrected = []
err_uncorrected = []

n = np.arange(5, 50)
for count in n:
    # Compute location and scale of (normal) null distribution
    # according to asymptotic approximation
    mn = count * (count + 1.) * 0.25
    se = count * (count + 1.) * (2. * count + 1.)
    se = np.sqrt(se / 24)

    # Exact null distribution
    X = WilcoxonDistribution(count)

    # Statistic values near typical significance threshold
    k = np.arange(X._dists[count].size)
    k = k[X.sf(k) < 0.2]

    # Uncorrected/Corrected null distribution
    Y = stats.norm(mn, se)
    Z = stats.norm(mn + 0.5, se)

    # Compute survival function (i.e. p-value with alternative='greater')
    p0 = X.sf(k)
    py = Y.sf(k)
    pz = Z.sf(k)

    # Record errors
    err_uncorrected.append(np.max(np.abs(py - p0)))
    err_corrected.append(np.max(np.abs(pz - p0)))

plt.plot(n, err_corrected, label='corrected')
plt.plot(n, err_uncorrected, label='uncorrected')
plt.xlabel('Sample size')
plt.ylabel('Maximum one-sided p-value disagreement (exact p < 0.2)')
plt.title('Error of asymptotic null distributions')
plt.legend()
```

</details>

Once the sample size is close to $n=50$, the maximum error of the corrected null distribution in this domain gets small enough that it is OK for most purposes to use the less expensive asymptotic calculation. (That said, the exact null distribution calculation is quite fast, so we might consider increasing the threshold even further.)

The stories will be the same for "greater" and "two-sided" p-values because the null distribution is symmetric.

chrisb83 commented:
> Perhaps instead we should only let "auto" resolve to a PermutationMethod if n is small enough to perform an exact permutation test (deterministic). Otherwise, emit a warning suggesting that the user specify a PermutationTest explicitly?

That is what I implemented now. The permutation test is only exact if the sample size is <= 13 (otherwise one has to increase n_resamples quite a bit and it is fairly slow. potentially one can speed it up using `batch`?) For n> 13, one could also switch to asymptotic instead of emitting a warning.

mdhaber commented on scipy/stats/_wilcoxon.py:
```diff
@@ -222,8 +213,26 @@ def _wilcoxon_nd(x, y=None, zero_method='wilcox', correction=True,
+                       "For sample size > 13, the p-value is simulated and "
+                       "not deterministic since `n_resamples` is capped at "
+                       "9999 to avoid long runtimes. Consider running "
+                       "wilcoxon using PermutationMethod with a larger "
```
```suggestion
                       "`wilcoxon` using a `PermutationMethod` with a larger "
```
and I think the first two sentences can be combined for concision, e.g.
> ... method=`auto` performs a randomized permutation test with 9999 resamples to limit execution time. Consider running...

mdhaber reviewed (commented):
Looking good. I think this currently implements the most substantial changes we might want to make with this PR. Defaulting to a randomized permutation test, even with a warning, may be even more aggressive than I'd propose. (What I had in mind was to use the asymptotic test but warn.) Let's see what other reviewers think.  I'll invite some specific maintainers to review in a bit if nobody jumps in.

In the meantime, consider comparing the "exact" (without considering ties), asymptotic, and *truly* exact (using permutation test) null distributions in an example case in which there are a few ties/zeros. For instance, when $n=20$ (which is probably still feasible to run offline), can a few ties/zeros make a substantial difference? What implications does it have for Type I/II errors? We don't have to tune this super tightly in this PR, but in the future it might be nice to study this more thoroughly and only run the permutation tests when there are enough ties to make a real difference.

chrisb83 commented:
> Looking good. I think this currently implements the most substantial changes we might want to make with this PR. Defaulting to a randomized permutation test, even with a warning, may be even more aggressive than I'd propose. (What I had in mind was to use the asymptotic test but warn.) Let's see what other reviewers think. I'll invite some specific maintainers to review in a bit if nobody jumps in.

OK, I misunderstood your suggestion. I think I will switch to asymptotic if the sample size is > 13 and if there are ties/zeros. I did not experiments and the accuracy is fine (I will add the details in a new comment). In that case, we also limit the impact on the output compared to the current behavior: we only do a permutation test if the results are deterministic, otherwise we keep the asymptotic method.

chrisb83 commented:
**Effect of ties**

- simulate difference `d = x - y` with 14 observations from standard normal and create `n_ties in {1, 2, 3, 5, 7}`
- compare p-values computed with `exact`, `asymptotic` and `PermutationMethod(n_resamples=np.inf)` (50 samples for each value of n_ties, two-sided test, `correction=True`)

Conclusion:
- The values computed with asymptotic and exact are very close to the ones computed with `PermutationMethod`.
- The results suggest that it is preferable to use asymptotic instead of exact if there are ties


Table showing stats of absolute differences of p-values:

<img width="413" alt="image" src="https://github.com/user-attachments/assets/ab9994a9-30db-4791-ac15-6db9a653b6f9">

Direct comparison of the sorted p-values depending on the number of ties:

![image](https://github.com/user-attachments/assets/26d316d4-d4a1-4ed3-bc98-c0366907c362)

![image](https://github.com/user-attachments/assets/826367fe-740e-4bb1-8c1d-19cacd8c64f5)

![image](https://github.com/user-attachments/assets/bf86540e-8d81-4757-b66b-c250a6abf5f5)

<details>

```python
from scipy import stats
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import time
rng = np.random.default_rng(789201128591203)
kwargs = {'alternative': "two-sided"}

n_ties_lst = [1, 2, 3, 5, 7]
n_test = 50

res_lst = []

tic = time.time()

for n_ties in n_ties_lst:
    print(n_ties, time.time() - tic)
    res = np.zeros((n_test, 3))
    for i in range(n_test):
        d = rng.standard_normal(size=14)
        d[1:(n_ties+1)] = d[0]
        res_pm = stats.wilcoxon(d, method=stats.PermutationMethod(n_resamples=np.inf), **kwargs)
        res_ex = stats.wilcoxon(d, method="exact", **kwargs)
        res_asymp = stats.wilcoxon(d, method="approx", correction=True, **kwargs)
        res[i, :] = (res_pm.pvalue, res_ex.pvalue, res_asymp.pvalue)
    df = pd.DataFrame(res)
    df.columns = ["p_permutation", "p_exact", "p_asymp"]
    df["n_ties"] = n_ties
    res_lst.append(df)

df = pd.concat(res_lst)
df.to_csv("tie_experiment.csv")
df.sort_values(by=["n_ties", "p_permutation"], inplace=True)

x = np.arange(n_test)
for i, n_ties in enumerate(n_ties_lst):
    fig, ax = plt.subplots()
    idx = df["n_ties"] == n_ties
    ax.plot(x, df.loc[idx, "p_permutation"], "x", label="permutation")
    ax.plot(x, df.loc[idx, "p_exact"], "x", label="exact")
    ax.plot(x, df.loc[idx, "p_asymp"], "x", label="asymp")
    ax.legend()
    ax.set_title(f"n_ties={n_ties}")
    plt.show()

for m in ["exact", "asymp"]:
    df[f"delta_{m}_vs_permutation"] = np.abs(df[f"p_{m}"] - df["p_permutation"])

d_cols = ["delta_exact_vs_permutation", "delta_asymp_vs_permutation"]
df.groupby("n_ties")[d_cols].agg(
    ["median", "mean", "std", "max"]
)
```
</details>

chrisb83 commented:
**Effect of zeros**

- simulate difference `d = x - y` with 14 observations from standard normal and create `n_ties in {1, 2, 3, 5, 7}`
- compare p-values computed with `exact`, `asymptotic` and `PermutationMethod(n_resamples=np.inf)` (50 samples for each value of n_ties, two-sided test, `correction=True, zero_method='zsplit'`)

Note that if `zero_method='wilcoxon'`, there is no need to use a permutation test since the zeros are dropped anyways. So using `method='exact'` will give the same p-value as a permutation test that iterates through all combinations. That is something we can consider in the implementation

Conclusion:

- The values computed with asymptotic and exact are generally close to the ones computed with PermutationMethod.
- The results suggest that it is preferable to use asymptotic instead of exact if there are zeros, though there is no big difference

Table showing stats of absolute differences of p-values:

<img width="410" alt="image" src="https://github.com/user-attachments/assets/090d9c6a-942f-4362-bc15-e4af04a61174">

Direct comparison of the sorted p-values depending on the number of zeros:

![image](https://github.com/user-attachments/assets/f96bd168-580b-430a-9e41-3bf827b43bc7)

![image](https://github.com/user-attachments/assets/9d146d4d-4eb3-4091-82c8-a4580ddf6e34)

![image](https://github.com/user-attachments/assets/f93d87ab-551a-4d41-a1eb-3ff93a3d64c5)

![image](https://github.com/user-attachments/assets/a0633953-5d97-4f46-8f1b-2b7a29eed1d9)

<details>

```python
from scipy import stats
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import time


rng = np.random.default_rng(1898734200237526)
kwargs = {'alternative': "two-sided", "zero_method": "zsplit"}

n_zeros_lst = [1, 2, 3, 5, 7]
n_test = 50

res_lst = []

tic = time.time()

for n_zeros in n_zeros_lst:
    print(n_zeros, time.time() - tic)
    res = np.zeros((n_test, 3))
    for i in range(n_test):
        d = rng.standard_normal(size=14)
        d[:n_zeros] = 0
        res_pm = stats.wilcoxon(d, method=stats.PermutationMethod(n_resamples=np.inf), **kwargs)
        res_ex = stats.wilcoxon(d, method="exact", **kwargs)
        res_asymp = stats.wilcoxon(d, method="approx", correction=True, **kwargs)
        res[i, :] = (res_pm.pvalue, res_ex.pvalue, res_asymp.pvalue)
    df = pd.DataFrame(res)
    df.columns = ["p_permutation", "p_exact", "p_asymp"]
    df["n_zeros"] = n_zeros
    res_lst.append(df)

df = pd.concat(res_lst)
df.to_csv("zeros_experiment.csv")
df.sort_values(by=["n_zeros", "p_permutation"], inplace=True)


x = np.arange(n_test)
for i, n_zeros in enumerate(n_zeros_lst):
    fig, ax = plt.subplots()
    idx = df["n_zeros"] == n_zeros
    ax.plot(x, df.loc[idx, "p_permutation"], "x", label="permutation")
    ax.plot(x, df.loc[idx, "p_exact"], "x", label="exact")
    ax.plot(x, df.loc[idx, "p_asymp"], "x", label="asymp")
    ax.legend()
    ax.set_title(f"n_zeros={n_zeros}")
    plt.show()


for m in ["exact", "asymp"]:
    df[f"delta_{m}_vs_permutation"] = np.abs(df[f"p_{m}"] - df["p_permutation"])

d_cols = ["delta_exact_vs_permutation", "delta_asymp_vs_permutation"]
df.groupby("n_zeros")[d_cols].agg(
    ["median", "mean", "std", "max"]
)

```
</details>

mdhaber commented:
@chrisb83 I'd like to ask for other feedback, but before I do, this should be in a state that we're happy with. Can you fix the test failures and address https://github.com/scipy/scipy/pull/21567#pullrequestreview-2321057562 / https://github.com/scipy/scipy/pull/21567#issuecomment-2381250241?

chrisb83 committed:
TST, DOC, MAINT: fix test failures, minor doc/code improvements in stats.wilcoxon

chrisb83 commented:
Hi Matt,

I addressed your comments and the test failures are hopefully resolved now. Main changes:

- use asymptotic if there are ties and sample size is too large for a deterministic permutation test with 9999 resamples (note that this was a bit trickier than I thought: first of all, there is another input validation that needs to be done once we switch to asymptotic which happens only after we compute the statistic)
- I also noted a small mistake in the current code: we always compute z although it is not needed unless method is asymptotic. it can actually cause a division by zero warning, e.g., if method is exact and there are only zeros in d)

The only thing I did not change are the computed values in the docstring. I will do that once we get further feedback

chrisb83 committed:
fix check that se = 0

chrisb83 commented on scipy/stats/_wilcoxon.py:
```diff
@@ -97,38 +97,35 @@ def _wilcoxon_iv(x, y, zero_method, correction, alternative, method, axis):
                       stacklevel=2)
 
-    if (method == "approx" and zero_method in ["wilcox", "pratt"]
+    if (method == "asymptotic" and zero_method in ["wilcox", "pratt"]
```
Does this condition make sense? If we use asymptotic with `d = np.zeros(10)`, we raise an error. If `d = np.zeros((10, 10))`,  one gets an array with statistic = 0, pvalue = nan.

mdhaber replied:
I don't remember, but I might have done it this way when I added native `axis` support to preserve existing behavior for 1-D arrays but not let a single slice of zeros in an N-D array cause the whole calculation to error out. I would just assume that we always return NaN - an appropriate result when there is not a numerical result - and document this condition*if* it is so confusing to the user why they can't get a p-values for degenerate data.

chrisb83 replied:
I agree, so you suggest to remove this check? I think it would be consistent with the behavior of N-D arrays. There will be a warning indicating that the input is not good and we can document it.

mdhaber replied:
That sounds good if the warning you're referring to is the standard divide by zero warning (not custom), so no accommodations are made in the code.

chrisb83 replied:
yes, exactly. the warning emitted w/o any changes to our code is "RuntimeWarning: invalid value encountered in divide z = (r_plus - mn) / se"

chrisb83 commented on scipy/stats/_wilcoxon.py:
```diff
@@ -179,9 +178,19 @@ def _wilcoxon_statistic(d, zero_method='wilcox'):
+    #   the statistic is calculated
+    # - for method != "asymptotic", avoid division by zero warning
+    #   (z is not needed anyways)
+    if np.any(se) == 0:
```
this might need to change depending on what inputs we allow if ndim > 1

chrisb83 replied:
I am thinking to add `output_z` as a keyword to the function that computes the statistic. It don't like the fact that you get a warning due to a calculation that is not needed unless `output_z = True`. So essentially the code would be:

```
if output_z:
     z = (r_plus - mn) / se
else:
   z = np.nan # no need to consider the dimension of se since the value is unused
```

wdyt?

mdhaber replied:
It needs to be calculated whenever the asymptotic method is used, and whether it is returned is already controlled by using `method='asymototic'`. SoI don't see a need for an additional keyword. 

I am not bothered by a warning one might get in an exceptional case. How did this come up - do you observe the warning frequently, happen to encounter it in the thousands of simulations you did with many ties, or did you see division by a variable in the code? If no user has complained about this in the past, I think there are other things we can do that will have greater impact. 

I'd like to keep the PR's scope to the four improvements, since it's complicated enough with those. If you consider this important enough to come back to separately, I'd be happy to discuss it then. Does that sound ok?

chrisb83 replied:
the calculation happens within `_wilcoxon_statistic(d, zero_method='wilcox')`, so the method is not a keyword

the reason why I consider the warning relevant for this PR is as follows: I noted that we raise a value error if all entries are zero in a 1D array in input validation. since we plan to switch to asymptotic in some cases _after_ the validation step, I had to repeat this check. And then I noted that the behavior is not consistent with n-d arrays. I think we agree on the point that we can skip the additional validation and just return NaN and a warning. However, my view is that we should only show the warning if it is relevant (for example, if method="auto" and the sample size is 10, we switch to PermutationMethod and z is not needed/used, and it works fine with d = 0, so we should not warn). so for me, this change goes together with the change of the input validation. I hope it clarifies.

mdhaber replied:
I misunderstood when you wrote:

> I am thinking to add output_z as a keyword to the function that computes the statistic

I didn't notice that you wrote that you would add it to the private function, not the public function. I thought you were proposing a new public argument.

If the change is very simple, OK.

chrisb83 committed:
MAINT, TST: handle edge case of d=0 in wilcoxon

mdhaber commented on scipy/stats/tests/test_morestats.py:
```diff
@@ -1476,13 +1476,17 @@ def test_wilcoxon_bad_arg(self):
-        assert_raises(ValueError, stats.wilcoxon, x, x, "pratt",
-                      mode="approx")
+        # pratt and wilcox do not work if x - y == 0 and method == "asymptotic"
+        # => warning is emitted and p-value is nan
+        msg = "invalid value encountered in scalar divide"
+        with pytest.warns(RuntimeWarning, match=msg):
+            w, p = stats.wilcoxon(x, x, "wilcox", method="asymptotic")
+        assert_equal((w, p), (0.0, np.nan))
+        with pytest.warns(RuntimeWarning, match=msg):
+            w, p = stats.wilcoxon(x, x, "pratt", method="asymptotic")
+        assert_equal((w, p), (0.0, np.nan))
```
```suggestion
        # => warning may be emitted and p-value is nan
        with np.errstate(invalid='ignore'):
            w, p = stats.wilcoxon(x, x, "wilcox", method="asymptotic")
            assert_equal((w, p), (0.0, np.nan))
            w, p = stats.wilcoxon(x, x, "pratt", method="asymptotic")
            assert_equal((w, p), (0.0, np.nan))
```

mdhaber commented:
@josef-pkt @raphaelvallat I'd like to get your opinion on this PR because it would change some p-values returned by `wilcoxon` (arguably for the better), and we haven't gotten any feedback on the [forum post](https://discuss.scientific-python.org/t/improvements-of-wilcoxon-test-backwards-compatability/1442).

The current default for the continuity correction (`correction`) is `False`. This would immediately change the default to `True`, since it tends to make the asymptotic approximation more accurate. That said, whether overall "accuracy" is what is most important is debatable; see https://github.com/scipy/scipy/pull/21567#issuecomment-2365224206 and surrounding posts for some analysis. Would this be a good change to make? If so, is it OK to make the change immediately with the thought that it *improves* the p-value, or would it need a `FutureWarning` (i.e. we'd make the default `correction=None` and emit the warning unless it's specified explicitly)?

I think the rest of the changes are pretty safe, but you're welcome to comment on them, too.

mdhaber commented on scipy/stats/_morestats.py:
```diff
@@ -3679,6 +3686,12 @@ def wilcoxon(x, y=None, zero_method="wilcox", correction=False,
     WilcoxonResult(statistic=6.0, pvalue=0.4375)
 
     """
+    # replace approx by asymptotic to ensure backwards compatability
```
I know we discussed this, but I just remembered that we allow both `mode` and `method` for this argument, so there would be some precedent if we want to just accept "approx" as an alias rather than being disruptive.

mdhaber commented on scipy/stats/_wilcoxon.py:
```diff
@@ -97,41 +97,33 @@ def _wilcoxon_iv(x, y, zero_method, correction, alternative, method, axis):
+
     if n_zero > 0 and method == "exact":
-        method = "approx"
         warnings.warn("Exact p-value calculation does not work if there are "
```
```suggestion
        warnings.warn("The returned p-value is not exact when there are "
```

mdhaber reviewed (approved):
There are a few things I might adjust/refactor on my own later (e.g. maybe I'll just remove the z calculation from `_wilcoxon_statistic` and put it where it's needed), but it looks like this does what it set out to do:
- Change the name of `method='approx'` to `method='asymptotic'`
- Change the default `correction=False` to `correction=True`
- Improve the `method='auto'` logic
- Respect the user's selection if they choose `method='exact'`

It also avoids an edge case inconsistency (formerly error for 1D input with all zeros but not for ND) and avoids unnecessarily computing the z-statistic when it won't be used.

Before merging, I'll wait for concensus about whether the change in default `correction` is OK, though.

raphaelvallat commented:
Hi @mdhaber,

Changing the default `correction` to True sounds good to me 👍 Thanks for asking though, appreciate it!

Raphael

mdhaber commented:
@josef-pkt do you have a moment to weigh in on the change from `correction=False` to `correction=True` in `wilcoxon`? Thanks!

josef-pkt commented:
I did not look at the details. I have not read literature on this. So I have no opinion on continuity correction here.

In general I don't use continuity correction in statsmodels (not at all or not by default).

example testing comparing two proportions (where I know the literature quite well)
using continuity correction (in the appropriate test) makes the normal test close to Fisher exact test.

However, Fisher exact test is conservative. The opposing literature recommends to match size on average even if the test over reject for some part of the sample space. So, in statsmodels I try to default to those methods.
"better" in the sense of being closer to an "exact" test is not necessarily what we want.

I don't know if that also applies here.

(in general, I'm very reluctant to change default methods in statsmodels for popular statistics or hypothesis test out of backwards compatibility or replication considerations. Although, in some cases I should have switched brute force before it got locked in by usage.)

mdhaber commented:
Alright. @chrisb83, let's not change the default now. Let's change that back and we can merge the rest.

josef-pkt commented:
switching the default is still possible later.
But there needs to be a clear advantage in small sample size and power. (still with conflict of always preserving size versus size on average)

2x2 binomial in small samples is an extreme case where discreteness matters and "exact" methods are usually conservative.
I so it mentioned in the literature for other case that conservativeness because of the discrete sample space is reduced when averaging over more variables smooths the discreteness. (I don't remember in which context, likely multinomial).
On the other hand, it's likely that continuity correction has little effect in those cases.

chrisb83 committed:
change continuity correcction back to False in wilcoxon

chrisb83 committed:
accpet approx in wilcoxon w/o deprecation warning

chrisb83 committed:
fix test failure

chrisb83 committed:
mark one test as xslow, fix refguide failure

chrisb83 commented:
There is one test failure:

```
___________________________ TestWilcoxon.test_auto ____________________________
[gw0] win32 -- Python 3.12.7 C:\hostedtoolcache\windows\Python\3.12.7\x64\python.exe
Test passed but took too long to run: Duration 2.0124163999998927s > 1.0s
```

The test has become slower since we run two permutation tests with n=13 which is slow. I could change it as follows:
in `test_auto`, run a test with a small sample, say n=4. then add a test that I mark "slow" for the edge case n=13. @mdhaber is it ok for you?

To fix the ref guide check, I will add `method='exact'`, otherwise we switch from exact to permutation test if there are ties and then we observe a differene in the results both because of the method changes and because the input changes due to the roundoff error...

mdhaber commented:
> then we observe a differene in the results both because of the method changes and because the input changes due to the roundoff error...

Hmm I'm not sure if we'd want to show `method='exact'` being used when there are ties when the p-value with the default permutation method would be more accurate. Why not just change the expected `0.4375` to `0.5`? I would suggest that the method changing is not really part of the cause because when there are no ties, the permutation method is equivalent to `method='exact'`. So the change in value is still solely due to the roundoff error.

mdhaber commented:
Thanks @chrisb83! I'm going to submit a follow-up PR to suggest some slight tweaks, but let's get this in.
